vmware-base/src/com/cloud/hypervisor/vmware/mo/DistributedVirtualSwitchMO.java

```diff
@@ -36,24 +37,67 @@
 public class DistributedVirtualSwitchMO extends BaseMO {
     @SuppressWarnings("unused")
     private static final Logger s_logger = Logger.getLogger(DistributedVirtualSwitchMO.class);
+    private static ConcurrentHashMap<String, List<String>> s_dvPortGroupCacheMap = null;
 
     public DistributedVirtualSwitchMO(VmwareContext context, ManagedObjectReference morDvs) {
         super(context, morDvs);
+        s_dvPortGroupCacheMap = new ConcurrentHashMap<String, List<String>>();
     }
 
     public DistributedVirtualSwitchMO(VmwareContext context, String morType, String morValue) {
         super(context, morType, morValue);
+        s_dvPortGroupCacheMap = new ConcurrentHashMap<String, List<String>>();
     }
 
     public void createDVPortGroup(DVPortgroupConfigSpec dvPortGroupSpec) throws Exception {
         List<DVPortgroupConfigSpec> dvPortGroupSpecArray = new ArrayList<DVPortgroupConfigSpec>();
         dvPortGroupSpecArray.add(dvPortGroupSpec);
-        _context.getService().addDVPortgroupTask(_mor, dvPortGroupSpecArray);
+        boolean dvPortGroupExists = false;
+        String dvSwitchInstance = _mor.getValue();
+        String dvPortGroupName = dvPortGroupSpec.getName();
+        String uniquedvPortGroupPerDvs = dvSwitchInstance + dvPortGroupName;
+        List<String> dvPortGroupList = null;
+        synchronized (uniquedvPortGroupPerDvs.intern()) {
+            // Looking up local cache rather than firing another API call to see if dvPortGroup exists already.
+            if (s_dvPortGroupCacheMap.containsKey(dvSwitchInstance)) {
+                dvPortGroupList = s_dvPortGroupCacheMap.get(dvSwitchInstance);
+                if (dvPortGroupList != null && dvPortGroupList.contains(dvPortGroupName)) {
+                    dvPortGroupExists = true;
+                }
+            }
+            if (!dvPortGroupExists) {
+                ManagedObjectReference task = _context.getService().addDVPortgroupTask(_mor, dvPortGroupSpecArray);
+                if (!_context.getVimClient().waitForTask(task)) {
+                    throw new Exception("Failed to create dvPortGroup " + dvPortGroupSpec.getName());
+                } else {
+                    if (s_dvPortGroupCacheMap.containsKey(dvSwitchInstance)) {
+                        dvPortGroupList = s_dvPortGroupCacheMap.get(dvSwitchInstance);
+                        if (dvPortGroupList == null) {
+                            dvPortGroupList = new ArrayList<String>();
+                        }
+                        dvPortGroupList.add(dvPortGroupName); //does this update map?
```

Comment:
Yes @sureshanaparti, you are right the `dvPortGroupList` ref would take care of the update.